tools/dds/py-dds-server/python.cpp

```diff
@@ -209,9 +215,67 @@ PYBIND11_MODULE(NAME, m) {
               } )
         .def( "create_listener", []( dds_participant & self ) { return self.create_listener(); } );
 
+    // The 'types' submodule will contain all the known topic types in librs
+    // These are used alongside the 'dds_topic' class:
+    //      topic = dds.topic( p, dds.types.device_info(), "realsense/device-info" )
+    auto types = m.def_submodule( "types", "all the types that " SNAME " can work with" );
+
+    // We need to declare a basic TypeSupport or else dds_topic ctor won't work
+    using eprosima::fastdds::dds::TypeSupport;
+    py::class_< TypeSupport > topic_type( types, "topic_type" );
+
+    // Another submodule, 'raw' is used with dds_topic to actually send and receive messages, e.g.:
+    //      reader = dds.topic_reader( topic )
+    //      ...
+    //      info = dds.device_info( dds.raw.device_info( reader ) )
+    //auto raw = m.def_submodule( "raw", "all the topics that " SNAME " can work with" );
+
+
+    using librealsense::dds::dds_topic;
+    py::class_< dds_topic, std::shared_ptr< dds_topic > >( m, "topic" )
+        .def( py::init< std::shared_ptr< dds_participant > const &,
+                        eprosima::fastdds::dds::TypeSupport const &,
+                        char const * >() )
+        .def( "get_participant", &dds_topic::get_participant )
+        .def( "__repr__", []( dds_topic const & self ) {
+            std::ostringstream os;
+            os << "<" SNAME ".topic \"" << self->get_name() << "\"";
+            os << ">";
+            return os.str();
+        } );
+
+    using librealsense::dds::dds_topic_reader;
+    py::class_< dds_topic_reader, std::shared_ptr< dds_topic_reader > >( m, "topic_reader" )
+        .def( py::init< std::shared_ptr< dds_topic > const & >() )
+        .def( FN_FWD( dds_topic_reader, on_data_available, ( ), ( ), callback(); ) )
+        .def( FN_FWD( dds_topic_reader,
+                      on_subscription_matched,
+                      (),
+                      (eprosima::fastdds::dds::SubscriptionMatchedStatus const &),
+                      callback(); ) )
+        .def( "run", &dds_topic_reader::run );
+
+
+    // The actual types are declared as functions and not classes: the py::init<> inheritance rules are pretty strict
+    // and, coupled with shared_ptr usage, are very hard to get around. This is much simpler...
     using librealsense::dds::topics::device_info;
+    using raw_device_info = librealsense::dds::topics::raw::device_info;
+    types.def( "device_info", []() { return TypeSupport( new device_info::type ); } );
+
     py::class_< device_info >( m, "device_info" )
         .def( py::init<>() )
+        //.def( py::init< raw_device_info const & >() )
+        .def( py::init( []( dds_topic_reader const & reader ) {
+            auto actual_type = reader.topic()->get()->get_type_name();
+            if( actual_type != "librealsense::dds::topics::raw::device_info" )
```

Comment:
Can't we take it somehow from out `device-info` topic header? `construct_topic_name`?

Reply:
This checks that you haven't mixed up a reader and a type -- it can crash because the reader itself does no type-checking.
I.e., you can have a small data structure and read "it" from a reader for a big data-structure, and the reader won't care.

Reply:
I understand but the reason I ask is that it's more safe to have a single place when we hard-coded this topic name.
If we decide in the future to change it, this usage can be missed.
If we were able to use the same source for it, it will be safer..

Reply:
The only place the **type** is mentioned is in the auto-generated PubSubType code. If I made another define in `device-info-msg.h` you'd have the same problem.

How do you suggest we go about this?

Reply:
I can add a function in `device_info` that does it, similar to what I suggested to Ohad in his CR:

```
device_info device_info::take( std::shared_ptr< dds_topic_reader > const & reader )
{
    ...
}
```

Reply:
I mixed topic name and topic type, topic type is harder to get I agree.
I think I like you proposal..:)

Reply:
This `take` function will copy the data. OK for `device_info` not for images.

Reply:
> I think I like you proposal..:)

See new commit